enhancements/olm/simplify-apis.md

```diff
@@ -0,0 +1,523 @@
+**Mitigation:** Make operator get and list privileged operations. `ClusterRoles` can also be created which apply only to specific resources by name.
+
+## Design Details
+
+### Test Plan
+
+Any e2e tests will live in OLM's repo as per convention and will be executed separately from openshift's e2e tests.
+
+The following features should have distinct e2e tests:
+
+- Component selection
+- Component status conditions
+- Generated resource label propagation
+- `Subscription` adoption
+- Install from `CatalogSource`
+- Install from bundle image
+- Direct `ClusterServiceVersion` application
+- Copied `ClusterServiceVersion` cleanup
+
+### Graduation Criteria
+
+#### Dev Preview -> Tech Preview
+
+- Users can track operator components with the `Operator` resource
+- Users can install an operator from a `CatalogSource` and a bundle image
+- `ClusterServiceVersions` can be installed without ancillary setup
+- Sufficient test coverage
+- End user documentation, relative API stability
+- v2alpha1 is bumped to v2beta1
+- Gather feedback from users rather than just developers
+
+#### Tech Preview -> GA
+
+- v2beta1 is bumped to v2
+- Existing copied `ClusterServiceVersions` on a cluster are cleaned up
+- More testing (upgrade, downgrade, scale)
+- Sufficient time for feedback
+- Feature flags removed
+- Available by default
+
+### Upgrade / Downgrade Strategy
+
+On upgrade, an existing cluster can continue to use OLM as it had before. OLM will infer and generate any new resources required.
+
+Any downgrade procedure would need to take into account that using many different install `ServiceAccounts` in a single namesace is not possible in the previous version of OLM. This means that all `Subscriptions` in a namespace would need to share an `OperatorGroup` with a `ServiceAccount` to drive installation of their resources. This may require that some `Subscriptions` are moved to new namespaces since they could have different scoping requirements. Additionally, the fact that `OperatorGroups` don't support specifying the namespace of a `SerivceAccount` must be taken into account.
+
+### Version Skew Strategy
+
+The API being added is net-new, and only addative changes are being made to existing OLM resources. Handling skew for upgrades/downgrades is discussed in [that section](#upgrade-/-downgrade-strategy).
+
+## Implementation History
+
+N/A
+
+## Drawbacks
+
+- A new API resource is required
+- Listing operators becomes a privileged operation that leaks info about existing namespaces and resources
+- Implict agreement that an operator is a cluster singleton
+
+## Alternatives
+
+We can define a namespace-scoped operator resource in addition to the cluster-scoped one. We can use the cluster-scoped resource to indicate an operator may span multiple namespaces, and a namespace scoped operator to indicate it should be installed only in a specific namespace.
```

Comment:
What are the use cases for operators that "span multiple namespaces"?  An operator can watch resources across the cluster with sufficient RBAC, so it's not about what it can touch.  And you could use multiple instances of `Operator` to install an operator more than once.

Reply:
> What are the use cases for operators that "span multiple namespaces"? An operator can watch resources across the cluster with sufficient RBAC, so it's not about what it can touch. 

We're trying to set the stage for a better operator author user story. An operator author knows best how their operator works, and it's come into question whether enforcing the use of a single namespace is a reasonable constraint. For example, some operators are written with hardcoded namespace, and others with `RoleBindings` in `kube-system`. Rather than "what it can touch", I'd say it's a "native configuration" story. Along with this we also need to add knobs for admins to map those native namespaces.

> And you could use multiple instances of Operator to install an operator more than once.

I don't think that multiple instances of an operator is a tenable configuration for a cluster. This comes down to the fact that conversion webhooks and extension api-servers are effectively singletons and require exactly one backing service to operate. While we can come up with creative (and probably complex) ways of intercepting and routing traffic, I think that any real solution will require better first-class multitenancy in kubernetes.

Reply:
When I think of the operator installation I think mostly about the `Deployment` and associated `Roles` and `RoleBindings` (permissions in CSV currently).  Anything outside of this is managed by part of the operator's bundle or managed by the operator's controller directly.  I see bundled and controller managed resources as outside of the scope of what `Operator` will manage.  It is possible today to create a bundle with `Roles` and `RoleBindings` outside of the namespace in which the `Deployment` is created, but `Subscription` is namespaced.  What is changing about the operator story that changes the scope of operator installation to imply operators are a cluster scoped resource?  At the end of the day, the core of an operator is the controller deployed as a pod by a deployment, which is namespaced.

Reply:
I don't think the operator story is changing; the implementation is evolving to better match the reality of the platform. As @njhale points out, having multiple instances of the same controller running concurrently is not a good fit for the way k8s is evolving its CRD story. As I understand it, the outcome is that you'll have one instance of a controller running in the cluster, but then you might be able to choose whether that controller watches one namespace, a list of namespaces, or all namespaces.